### PR TITLE
fix: Backport opentelemetry 0.18 code from main repo

### LINKED ISSUE #8: Update opentelemetry to 0.18
## Feature Request
Update opentelemetry to 0.18

### Motivation
It seems that 0.18 broke a bunch of apis, so it seems it deserves a dedicated issue.

### PR DESCRIPTION
**Pull request by jaysonsantos**
## Motivation
I noticed that this repo got the cut-off before opentelemetry 0.18 was committed into the main repo so I tried to backport what was there.

## Solution
This is pretty much a backport from the code that @jtescher did on the main repo.

Let me know if you see something wrong, I tried to adapt the rejections the best way possible.

This fix #8

### CODE REVIEW

**Review by jaysonsantos — COMMENTED** [review]
I just left comments on parts of the code that seems that the current version should stay, would you mind taking a look at them?

**Comment by djc** [comment]
Note that opentelemetry 0.19 has been released yesterday.

**Comment by jaysonsantos** [comment]
@djc I've tested 0.19 with this PR, and it seems to work fine but, as this is a backport from another author, I didn't want to change the code in this PR as I don't “own” it

**Comment by davidbarsky** [comment]
Thanks for doing this PR! Sorry for the delay in reviewing this.

>  I've tested 0.19 with this PR, and it seems to work fine but, as this is a backport from another author, I didn't want to change the code in this PR as I don't “own” it

I think that for clarity of commit history, we should make upgrading to opentelemetry 0.19 a separate PR.

File: src/tracer.rs [hunk 1]
```diff
@@ -74,19 +75,18 @@ impl PreSampledTracer for Tracer {
...
         // Gather trace state
-        let (no_parent, trace_id, remote_parent, parent_trace_flags) =
-            current_trace_state(builder, parent_cx, &provider);
+        let (trace_id, parent_trace_flags) = current_trace_state(builder, parent_cx, &provider);
 
         // Sample or defer to existing sampling decisions
         let (flags, trace_state) = if let Some(result) = &builder.sampling_result {
             process_sampling_result(result, parent_trace_flags)
-        } else if no_parent || remote_parent {
+        } else {
```
[the remark this answers is not in the document]
Comment (jtescher) [reply src/tracer.rs:83]: no this is right

File: src/tracer.rs [hunk 2]
```diff
@@ -95,12 +95,6 @@ impl PreSampledTracer for Tracer {
                 builder.sampling_result.as_ref().unwrap(),
                 parent_trace_flags,
             )
-        } else {
```
[the remark this answers is not in the document]
Comment (jtescher) [reply src/tracer.rs:98]: This changed as well

**Review by jtescher — APPROVED** [review]

**Comment by RenanSFreitas** [comment]
Any chances this gets merged soon? I'd like to use some features present in opentelemetry 0.19.0 together with tracing-opentelemetry.

**Comment by shaun-cox** [comment]
@jaysonsantos, looks like `examples/opentelemetry.rs` needs to be updated as well to use `opentelemetry_jaeger::new_agent_pipeline()` instead of `opentelemetry_jaeger::new_pipeline()` in order to fix the above checks?

I'm eager to see this merged too (as well as the update for opentelemetry 0.19).  If there is anything I can do to help, please let me know...  Thanks for shepherding this!

**Comment by jaysonsantos** [comment]
@davidbarsky @jtescher do you mind taking a look on the latest commit?

**Comment by Dzejkop** [comment]
@davidbarsky @jtescher do you know when you'll be able to take a look at this? I'm wondering if I should wait or use a temporary fork

File: src/layer.rs [hunk 3]
```diff
@@ -1104,7 +1119,7 @@ mod tests {
         let subscriber = tracing_subscriber::registry().with(layer().with_tracer(tracer.clone()));
 
         tracing::subscriber::with_default(subscriber, || {
-            tracing::debug_span!("request", otel.kind = %SpanKind::Server);
+            tracing::debug_span!("request", otel.kind = "server");
```
[the remark this answers is not in the document]
Comment (jtescher) [reply src/layer.rs:1122]: The spec doesn't define a `Display` impl so this was removed in the `opentelemetry` crate as it pushes for a 1.0 release, so we have to provide our own conventions in this crate for now.

**Comment by jtescher** [comment]
Thanks @jaysonsantos!

**Pull request merged**